include/amqpcpp/libev.h

```diff
@@ -254,9 +254,11 @@ class LibEvHandler : public TcpHandler
             {
                 // the server was inactive for a too long period of time, reset state
                 _next = _expire = 0.0; _timeout = 0;
-                
+
+                ev_timer_stop(_loop, &_timer);
+
```

Comment:
I don't get this change. The fact that the timer expires implicitly means that it is stopped, so there shouldn't be a need to explicitly stop it, right?

Reply:
Setting _timer.repeat means that timer becomes repeating. Look at man ev(3):

>    "ev_timer" - relative and optionally repeating timeouts
>            Timer watchers are simple relative timers that generate an event after a given time, and optionally repeating in regular intervals after that.

Callback of repeating  timer is called regularly until ev_timer_stop() is called.

Please look at the test code.

Reply:
@EmielBruijntjes To see repeat behaviour you can run this simple test:
```c++
#include <iostream>
#include <ev.h>

static void
callback(struct ev_loop *loop, struct ev_timer *timer, int revents)
{
    std::cout << "callback" << std::endl;
    return;
}

int main()
{
    ev_timer timer;
    struct ev_loop *loop = EV_DEFAULT;

    ev_init(&timer, callback);
    timer.repeat = 3;
    ev_timer_start(loop, &timer);
    ev_run(loop, 0);

    return 0;
}
```

Reply:
Yes you are right, I forgot about that repeat-setting, I will check it out a bit later

Reply:
I think better fix is to call ev_timer_stop() unconditionally in ~Wrapper(), i.e. do not check timed(). Man says:

>  "ev_TYPE_stop" (loop, ev_TYPE *watcher)
         ...
         If you want to free or reuse the memory used by the watcher it
         is therefore a good idea to always call its "ev_TYPE_stop" function.

However it seems the fix should contain ev_ref/ev_unref calls as well. Could you point what should be achieved in the end by using those? I see ev_timer_again() call is followed by ev_unref() in onExpired(). But ev_timer_again() call in Wrapper::start() is not followed. So I am confused.

Reply:
Sorry for being a little late. I made a mistake with setting the repeat property and the call to 'ev_timer_again()'. I did not realize that this would create a _repeating_ timer. Our code is written with the idea in mind that the timer expires _once_, and then is re-armed for the _next_ timerout. I thought that the ev_timer_again() function would only trigger the timer one more time. So this code was wrong:
```
// reset the timer to trigger again later
_timer.repeat = std::min(_next, _expire) - now;

// restart the timer
ev_timer_again(_loop, &_timer);
```
Should probably be replaced with this to get it to function the way I intended:
```
// reset the timer to trigger again later
ev_timer_set(&_timer, std::min(_next, _expire) - now, 0.0);

// restart the timer
ev_timer_start(_loop, &_timer);
```

Reply:
Could you be so helpful to adjust your pull request and see if this change fixes it too?

Reply:
I observe a similar segfault, on tagged version v4.1.6. I get a segfault on while attempting to recover from a disconnect event. To recover from a disconnect event, I instantiate a new `AMQP::TcpConnection`, and pass it to a new `AMQP::TcpChannel` object. Segfault follows shortly with a backtrace that looks like this: 
```
0: new_allocator<AMQP::Monitor*>::construct<AMQP::Monitor*, AMQP::Monitor* const&> at new_allocator.h:136
1: std::allocator_traits<std::allocator<AMQP::Monitor*> >::construct<AMQP::Monitor*, AMQP::Monitor* const&> at alloc_traits.h:475
2: std::vector<AMQP::Monitor*, std::allocator<AMQP::Monitor*> >::push_back ( at stl_vector.h:943
3:  AMQP::Watchable::add at AMQP-CPP/include/amqpcpp/watchable.h:50
4: AMQP::Monitor::Monitor at AMQP-CPP/include/amqpcpp/monitor.h:57
5: AMQP::TcpConnection::close at AMQP-CPP/src/linux_tcp/tcpconnection.cpp:115
6: AMQP::LibEvHandler::Wrapper::onExpired at AMQP-CPP/include/amqpcpp/libev.h:251
7: AMQP::LibEvHandler::Wrapper::callback at AMQP-CPP/include/amqpcpp/libev.h:214
8: ev_invoke_pending ()
9: ev_run ()
```

per @EmielBruijntjes  suggestion, the following diff indeed fixes the segfault
```
diff --git a/include/amqpcpp/libev.h b/include/amqpcpp/libev.h
index 110f251..0f61725 100644
--- a/include/amqpcpp/libev.h
+++ b/include/amqpcpp/libev.h
@@ -267,12 +267,12 @@ private:
                 // sent only after _timout/2 seconds again _from now_ (no catching up)
                 _next = now + std::max(_timeout / 2, 1);
             }
-            
+
             // reset the timer to trigger again later
-            _timer.repeat = std::min(_next, _expire) - now;
-            
+            ev_timer_set(&_timer, std::min(_next, _expire) - now, 0.0);
+
             // restart the timer
-            ev_timer_again(_loop, &_timer);
+            ev_timer_start(_loop, &_timer);
             
             // and because the timer is running again, we restore the refcounter
             ev_unref(_loop);
```
similar fix should be probably applied to `uint16_t start(uint16_t timeout)` method, but testing suggests that reconnect times increase.

When can we expect this bug to be fixed?
Thanks!

Reply:
> When can we expect this bug to be fixed?
> Thanks!

It must be fixed in #423